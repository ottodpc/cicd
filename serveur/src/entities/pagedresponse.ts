import { PageInfo } from './pageinfo'

export class PagedResponse {
	results: any[]

	pageInfo: PageInfo

	constructor(results: any[], pageInfo: PageInfo) {
		this.results = results
		this.pageInfo = pageInfo
	}
}
