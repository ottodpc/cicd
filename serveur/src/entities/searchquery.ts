export class SearchQuery {
	filterName: string

	operation: string

	filterValue: any

	constructor(filterName: string, operation: string, filterValue: any) {
		this.filterName = filterName
		this.operation = operation
		this.filterValue = filterValue
	}
}
