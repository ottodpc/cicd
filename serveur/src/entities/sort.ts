export class Sort {
	columnName: string

	orderType: string

	constructor(columnName: string, orderType: string) {
		this.columnName = columnName
		this.orderType = orderType
	}
}
