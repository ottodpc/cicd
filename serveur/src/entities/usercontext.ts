export class UserContext {
	userId: string

	constructor(userId: string) {
		this.userId = userId
	}
}
