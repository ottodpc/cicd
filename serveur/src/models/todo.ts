import { Schema } from '@nestjs/mongoose'

import { Prop } from '@nestjs/mongoose'
import { ApiProperty } from '@nestjs/swagger'
import { SchemaFactory } from '@nestjs/mongoose'

@Schema()
export class Todo {
	@Prop()
	@ApiProperty()
	title: string

	@Prop()
	@ApiProperty()
	description: string

	@Prop()
	@ApiProperty()
	todoId: string

	@Prop()
	_id: string
}

export const TodoSchema: SchemaFactory = SchemaFactory.createForClass(Todo)
