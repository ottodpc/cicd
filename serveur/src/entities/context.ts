import { UserContext } from './usercontext'

export class Context {
	userContext: UserContext

	isAuthenticated: boolean

	constructor(userContext: UserContext, isAuthenticated: boolean) {
		this.userContext = userContext
		this.isAuthenticated = isAuthenticated
	}
}
