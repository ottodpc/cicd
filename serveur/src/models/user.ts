import { Schema } from '@nestjs/mongoose';

import { Prop } from '@nestjs/mongoose';
import { ApiProperty } from '@nestjs/swagger';
import { Todo } from './todo';
import { ValidateNested } from 'class-validator';
import { Type } from 'class-transformer';
import { SchemaFactory } from '@nestjs/mongoose';

@Schema()
export class User {
  @Prop()
  @ApiProperty()
  username: string;

  @Prop()
  @ApiProperty({ type: Todo, isArray: true })
  @ValidateNested({ each: true })
  @Type(() => Todo)
  todos: Todo[];

  @Prop()
  @ApiProperty()
  password: string[];

  @Prop()
  @ApiProperty()
  userId: string;

  @Prop()
  @ApiProperty()
  _id: string;
}

export const UserSchema: SchemaFactory = SchemaFactory.createForClass(User);
