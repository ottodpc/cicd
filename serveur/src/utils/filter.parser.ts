import { regex } from './find.all.constants'
import { fieldTypeMapper } from './find.all.constants'
import { isFilterOperationValid } from './validator'
import { ParserException } from '../exceptions/parser.exception'
import { SearchQuery } from '../entities/searchquery'

export const filterParser = (
	filters: string,
	modelName: string
): SearchQuery[] => {
	try {
		const searchQuery: SearchQuery[] = []
		const errorMessages: string[] = []
		if (filters === '') {
			return searchQuery
		}
		filters.split(':and:').forEach((filter) => {
			const matchResult = filter.match(regex)
			const filterName = matchResult[1]
			const operation = matchResult[2]
			const filterValue = getFieldValue(
				filterName,
				matchResult[3],
				modelName
			)
			if (isFilterOperationValid(filterName, operation, modelName)) {
				searchQuery.push(
					new SearchQuery(filterName, operation, filterValue)
				)
			} else {
				errorMessages.push(
					`operation ${operation} is not allowed on ${filterName}`
				)
			}
		})

		if (errorMessages.length > 0) {
			throw new ParserException(errorMessages.join(', \n'))
		}
		return searchQuery
	} catch (e) {
		throw new ParserException('failed to parse')
	}
}

const getFieldValue = (
	fieldName: string,
	fieldValue: any,
	modelName: string
) => {
	const modelTypeMapper = fieldTypeMapper.get(modelName) || new Map()
	const type = modelTypeMapper.get(fieldName) || ''

	if (type === 'string') {
		return fieldValue.toString()
	} else if (type === 'number') {
		return parseInt(fieldValue, 10)
	} else if (type === 'Date') {
		return new Date(fieldValue)
	} else if (type === 'Boolean') {
		return fieldValue.toString().toLowerCase() == 'true'
	}
	return fieldValue.toString()
}
