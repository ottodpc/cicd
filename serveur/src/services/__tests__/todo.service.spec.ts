import { TodoRepository } from '../../repositories/todo.repository'
import { TodoService } from '../todo.service'
import { Todo } from '../../models/todo'
import { readResourceFromExactPath } from '../../../test/utils'
import { CreateTodoInput } from '../../inputs/todo.input'
import { UserContext } from '../../entities/usercontext'
import { Context } from '../../entities/context'
import { UpdateTodoInput } from '../../inputs/todo.update'
describe('should test todoService', () => {
	let todoService: TodoService
	let todoRepository: TodoRepository

	beforeEach(async () => {
		todoRepository = {
			save: jest.fn(),
			findById: jest.fn(),
			findAll: jest.fn(),
			deleteById: jest.fn(),
		} as unknown as TodoRepository
		todoService = new TodoService(todoRepository)
	})

	it('should test create method', async () => {
		jest.spyOn(todoRepository, 'save').mockReturnValue(
			Promise.resolve(
				await JSON.parse(
					readResourceFromExactPath(
						__dirname + '/resources/Todo.json'
					)
				)
			)
		)
		const actualResp = await todoService.create(
			'defaultString',
			await JSON.parse(
				readResourceFromExactPath(
					__dirname + '/resources/createTodoInput.json'
				)
			),
			new Context(new UserContext('defaultString'), false)
		)
		expect(jest.spyOn(todoRepository, 'save')).toBeCalledTimes(1)
		expect(actualResp).toMatchSnapshot()
	})

	it('should test update method', async () => {
		jest.spyOn(todoRepository, 'save').mockReturnValue(
			Promise.resolve(
				await JSON.parse(
					readResourceFromExactPath(
						__dirname + '/resources/Todo.json'
					)
				)
			)
		)
		jest.spyOn(todoRepository, 'findById').mockReturnValue(
			Promise.resolve(
				await JSON.parse(
					readResourceFromExactPath(
						__dirname + '/resources/Todo.json'
					)
				)
			)
		)
		const actualResp = await todoService.update(
			'defaultString',
			await JSON.parse(
				readResourceFromExactPath(
					__dirname + '/resources/updateTodoInput.json'
				)
			),
			new Context(new UserContext('defaultString'), false)
		)
		expect(jest.spyOn(todoRepository, 'save')).toBeCalledTimes(1)
		expect(jest.spyOn(todoRepository, 'findById')).toBeCalledTimes(1)
		expect(actualResp).toMatchSnapshot()
	})

	it('should test get method', async () => {
		jest.spyOn(todoRepository, 'findById').mockReturnValue(
			Promise.resolve(
				await JSON.parse(
					readResourceFromExactPath(
						__dirname + '/resources/Todo.json'
					)
				)
			)
		)
		const actualResp = await todoService.get(
			'defaultString',
			new Context(new UserContext('defaultString'), false)
		)
		expect(jest.spyOn(todoRepository, 'findById')).toBeCalledTimes(1)
		expect(actualResp).toMatchSnapshot()
	})

	it('should test select method', async () => {
		jest.spyOn(todoRepository, 'findAll').mockReturnValue(
			Promise.resolve([
				await JSON.parse(
					readResourceFromExactPath(
						__dirname + '/resources/Todo.json'
					)
				),
			])
		)
		const actualResp = await todoService.select(
			'defaultString',
			'defaultString',
			123,
			123,
			new Context(new UserContext('defaultString'), false)
		)
		expect(jest.spyOn(todoRepository, 'findAll')).toBeCalledTimes(1)
		expect(actualResp).toMatchSnapshot()
	})

	it('should test delete method', async () => {
		jest.spyOn(todoRepository, 'findById').mockReturnValue(
			Promise.resolve(
				await JSON.parse(
					readResourceFromExactPath(
						__dirname + '/resources/Todo.json'
					)
				)
			)
		)
		jest.spyOn(todoRepository, 'deleteById').mockImplementation(undefined)
		const actualResp = await todoService.delete(
			'defaultString',
			new Context(new UserContext('defaultString'), false)
		)
		expect(jest.spyOn(todoRepository, 'findById')).toBeCalledTimes(1)
		expect(jest.spyOn(todoRepository, 'deleteById')).toBeCalledTimes(1)
		expect(actualResp).toMatchSnapshot()
	})
})
