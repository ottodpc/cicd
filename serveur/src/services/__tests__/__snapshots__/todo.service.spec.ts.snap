// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should test todoService should test create method 1`] = `
{
  "_id": "fugiat Lorem irure",
  "description": "fugiat Lorem irure",
  "fugiat00": -16000000,
  "title": "fugiat Lorem irure",
  "todoId": "fugiat Lorem irure",
}
`;

exports[`should test todoService should test delete method 1`] = `undefined`;

exports[`should test todoService should test get method 1`] = `
{
  "_id": "fugiat Lorem irure",
  "description": "fugiat Lorem irure",
  "fugiat00": -16000000,
  "title": "fugiat Lorem irure",
  "todoId": "fugiat Lorem irure",
}
`;

exports[`should test todoService should test select method 1`] = `
[
  {
    "_id": "fugiat Lorem irure",
    "description": "fugiat Lorem irure",
    "fugiat00": -16000000,
    "title": "fugiat Lorem irure",
    "todoId": "fugiat Lorem irure",
  },
]
`;

exports[`should test todoService should test update method 1`] = `
{
  "_id": "fugiat Lorem irure",
  "description": "fugiat Lorem irure",
  "fugiat00": -16000000,
  "title": "fugiat Lorem irure",
  "todoId": "fugiat Lorem irure",
}
`;
